contracts/BC_fusion/StakeHub.sol

```diff
@@ -0,0 +1,949 @@
+    using Utils for bytes;
+    using EnumerableSet for EnumerableSet.AddressSet;
+
+    /*----------------- constants -----------------*/
+    uint256 private constant BLS_PUBKEY_LENGTH = 48;
+    uint256 private constant BLS_SIG_LENGTH = 96;
+
+    address public constant DEAD_ADDRESS = address(0xdead);
+    uint256 public constant INIT_LOCK_AMOUNT = 1 ether;
+    uint256 public constant REDELEGATE_FEE_RATE_BASE = 10000; // 100%
+
+    //TODO
+    bytes private constant INIT_BC_CONSENSUS_ADDRESSES =
+        hex"00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000000";
+    bytes private constant INIT_BC_VOTE_ADDRESSES =
+        hex"00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000000";
+
+    /*----------------- errors -----------------*/
+    error StakeHubPaused();
+    error InBlackList();
+    error OnlyAssetProtector();
+    error ValidatorExisted();
+    error ValidatorNotExist();
+    error ValidatorNotJailed();
+    error DuplicateConsensusAddress();
+    error DuplicateVoteAddress();
+    error SelfDelegationNotEnough();
+    error InvalidCommission();
+    error InvalidMoniker();
+    error InvalidVoteAddress();
+    error InvalidConsensusAddress();
+    error UpdateTooFrequently();
+    error JailTimeNotExpired();
+    error DelegationAmountTooSmall();
+    error OnlySelfDelegation();
+    error ZeroShares();
+    error SameValidator();
+    error NoMoreFelonyToday();
+    error AlreadySlashed();
+    error TransferFailed();
+
+    /*----------------- storage -----------------*/
+    bool private _paused;
+    uint8 private _isRedelegating;
+    uint256 public transferGasLimit;
+
+    // stake params
+    uint256 public minSelfDelegationBNB;
+    uint256 public minDelegationBNBChange;
+    uint256 public maxElectedValidators;
+    uint256 public unbondPeriod;
+    uint256 public redelegateFeeRate;
+
+    // slash params
+    uint256 public downtimeSlashAmount;
+    uint256 public felonySlashAmount;
+    uint256 public downtimeJailTime;
+    uint256 public felonyJailTime;
+
+    // validator operator address set
+    EnumerableSet.AddressSet private _validatorSet;
+    // validator operator address => validator info
+    mapping(address => Validator) private _validators;
+    // validator vote address => validator operator address
+    mapping(bytes => address) private _voteToOperator;
+    // validator consensus address => validator operator address
+    mapping(address => address) private _consensusToOperator;
+    // slash key => slash jail time
+    mapping(bytes32 => uint256) private _felonyRecords;
+    // operator address => day index => receivedReward
+    mapping(address => mapping(uint256 => uint256)) private _rewardRecords;
+    // operator address => day index => totalPooledBNB
+    mapping(address => mapping(uint256 => uint256)) private _totalPooledBNBRecords;
```

Comment:
Let us move this to the credit contract.

Reply:
Done